oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java

```diff
@@ -0,0 +1,278 @@
+            Thread.currentThread().setName(originalName);
+        }
+    }
+
+    private void downloadRange(TraversingRange range) throws InterruptedException, TimeoutException {
+        BasicDBObject[] block = new BasicDBObject[batchSize];
+        int nextIndex = 0;
+        BsonDocument findQuery = range.getFindQuery();
+        LOG.info("Traversing from {}: query: {}", range, findQuery);
+        FindIterable<BasicDBObject> mongoIterable = dbCollection
+                .withReadPreference(readPreference)
+                .find(findQuery)
+                .sort(ascending(NodeDocument.MODIFIED_IN_SECS, NodeDocument.ID));
+        try (MongoCursor<BasicDBObject> cursor = mongoIterable.iterator()) {
+            while (cursor.hasNext()) {
+                BasicDBObject next = cursor.next();
+                String id = next.getString(Document.ID);
+                this.nextLastModified = next.getLong(NodeDocument.MODIFIED_IN_SECS);
+                this.lastIdDownloaded = id;
+                this.documentsRead++;
+                if (filter.test(id)) {
+                    block[nextIndex] = next;
+                    nextIndex++;
+                    if (nextIndex == batchSize) {
+                        tryEnqueue(block);
+                        block = new BasicDBObject[batchSize];
+                        nextIndex = 0;
+                    }
+                }
+            }
+        }
+        LOG.info("Finished downloading range: {}", range);
+        if (nextIndex > 0) {
+            LOG.info("Enqueueing last block of size: {}", nextIndex);
+            BasicDBObject[] lastBlock = new BasicDBObject[nextIndex];
+            System.arraycopy(block, 0, lastBlock, 0, nextIndex);
+            tryEnqueue(lastBlock);
+        }
+    }
+
+    private void tryEnqueue(BasicDBObject[] block) throws TimeoutException, InterruptedException {
+        if (!mongoDocQueue.offer(block, MONGO_QUEUE_OFFER_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS)) {
+            throw new TimeoutException("Timeout trying to enqueue batch of MongoDB documents. Waited " + MONGO_QUEUE_OFFER_TIMEOUT);
+        }
+    }
+
+    private long getFirstLastModified(MongoCollection<BasicDBObject> dbCollection) {
+        return getModified(dbCollection, ascending(NodeDocument.MODIFIED_IN_SECS));
+    }
+
+    private long getLatestLastModified(MongoCollection<BasicDBObject> dbCollection) {
+        return getModified(dbCollection, descending(NodeDocument.MODIFIED_IN_SECS));
+    }
+
+    private long getModified(MongoCollection<BasicDBObject> dbCollection, Bson sort) {
+        BsonDocument query = new BsonDocument();
+        query.append(NodeDocument.MODIFIED_IN_SECS, new BsonDocument().append("$ne", new BsonNull()));
+        try (MongoCursor<BasicDBObject> cursor = dbCollection
+                .find(query)
+                .sort(sort)
+                .limit(1)
+                .iterator()) {
+            if (!cursor.hasNext()) {
+                return -1;
```

Comment:
This should never happen, it would mean the collection is empty. What about throwing an exception in this case? In any case, if the collection is empty, there is nothing to index, so we can as well fail the job.